apollo-router/src/cache/size_estimation.rs

```diff
@@ -0,0 +1,438 @@
+/// Yes, it's inaccurate, but we're looking for something that is relatively cheap to compute.
+/// It doesn't take into account shared datastructures occurring multiple times and will give the
+/// full estimated serialized cost.
+#[derive(Default, Debug)]
+struct CountingSerializer {
+    count: usize,
+}
+
+impl ser::Serializer for CountingSerializer {
+    type Ok = Self;
+    type Error = Error;
+    type SerializeSeq = Self;
+    type SerializeTuple = Self;
+    type SerializeTupleStruct = Self;
+    type SerializeTupleVariant = Self;
+    type SerializeMap = Self;
+    type SerializeStruct = Self;
+    type SerializeStructVariant = Self;
+
+    fn serialize_bool(mut self, _v: bool) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<bool>();
+        Ok(self)
+    }
+
+    fn serialize_i8(mut self, _v: i8) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<i8>();
+        Ok(self)
+    }
+
+    fn serialize_i16(mut self, _v: i16) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<i16>();
+        Ok(self)
+    }
+
+    fn serialize_i32(mut self, _v: i32) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<i32>();
+        Ok(self)
+    }
+
+    fn serialize_i64(mut self, _v: i64) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<i64>();
+        Ok(self)
+    }
+
+    fn serialize_u8(mut self, _v: u8) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<u8>();
+        Ok(self)
+    }
+
+    fn serialize_u16(mut self, _v: u16) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<u8>();
+        Ok(self)
+    }
+
+    fn serialize_u32(mut self, _v: u32) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<u32>();
+        Ok(self)
+    }
+
+    fn serialize_u64(mut self, _v: u64) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<u64>();
+        Ok(self)
+    }
+
+    fn serialize_f32(mut self, _v: f32) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<f32>();
+        Ok(self)
+    }
+
+    fn serialize_f64(mut self, _v: f64) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<f64>();
+        Ok(self)
+    }
+
+    fn serialize_char(mut self, _v: char) -> Result<Self::Ok, Self::Error> {
+        self.count += std::mem::size_of::<char>();
+        Ok(self)
+    }
+
+    fn serialize_str(mut self, v: &str) -> Result<Self::Ok, Self::Error> {
+        //ptr + 8 bytes length + 8 bytes capacity
+        self.count += 24 + v.len();
+        Ok(self)
+    }
+
+    fn serialize_bytes(mut self, v: &[u8]) -> Result<Self::Ok, Self::Error> {
+        self.count += v.len();
+        Ok(self)
+    }
+
+    fn serialize_none(self) -> Result<Self::Ok, Self::Error> {
+        Ok(self)
```

Comment:
Just to double check: `None` would still use the size of `Some` in memory, right?
We estimate it as 0 because of serve limitations?

Reply:
Yeah, I guess in theory we can do something better. But None values won't have strings etc in them so I don't think we should worry about this and maybe we can follow up later to improve things.